code/datums/progress_bar.dm

```diff
@@ -0,0 +1,213 @@
+	var/mob/user
+	/// The client currently owning the progress bar for display.
+	var/client/client
+	/// The bar's index number within the owning mob's active progress bar list.
+	var/list_index
+	/// Whether the bar is stopping.
+	var/stopping = FALSE
+	/// Stored progress number for the goal fail effect.
+	var/last_progress = 0
+
+/datum/progress_bar/New(mob/param_user, param_goal, atom/param_target)
+	. = ..()
+	if(!istype(param_target))
+		throw EXCEPTION("datum/progress_bar: Invalid param given - param_target")
+
+	user = param_user
+	goal = param_goal
+	target = param_target
+	init_images(target)
+
+	if(user)
+		client = user.client
+		client?.images |= holder
+		LAZYASSOCLISTADD(user.progress_bars, target, src)
+		animate(holder, pixel_z = world.icon_size + (height * LAZYLEN(user.progress_bars[target]) - 1), alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+	else
+		animate(holder, alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+
+/datum/progress_bar/Destroy()
+	client?.images -= holder
+
+	QDEL_NULL(holder)
+	QDEL_NULL(backdrop)
+	QDEL_NULL(fill)
+
+	LAZYASSOCLISTREMOVE(user.progress_bars, target, src)
+
+	target = null
+	user = null
+	client = null
+
+	return ..()
+
+/**
+  * Stops the progress bar and destoy it after a bit.
+  */
+/datum/progress_bar/proc/stop()
+	set waitfor = FALSE
+
+	shift_down_all()
+	update(last_progress)
+	animate(holder, alpha = 0, time = animation_time_fade, flags = ANIMATION_PARALLEL)
+	LAZYASSOCLISTREMOVE(user.progress_bars, target, src)
+
+	sleep(animation_time_fade)
+	qdel(src)
+
+/**
+  * Calls the shift animation on all attached bars.
+  */
+/datum/progress_bar/proc/shift_down_all()
+	if(user?.progress_bars[target])
+		var/list/bars = user.progress_bars[target]
+		for(var/i = bars.Find(src), i <= length(bars), i++)
+			var/datum/progress_bar/P = bars[i]
+			P.shift()
+
+/**
+  * Initializes the visual elements.
+  *
+  * Arguments:
+  * * target - The atom to parent the images to.
+  */
+/datum/progress_bar/proc/init_images(target)
+	holder = image('icons/effects/effects.dmi', target, "nothing", HUD_ABOVE_ITEM_LAYER)
+	holder.mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	holder.alpha = 0
+	holder.plane = HUD_PLANE
+	holder.appearance_flags = KEEP_TOGETHER | APPEARANCE_UI_IGNORE_ALPHA
+
+	backdrop = image(icon, target, "backdrop")
+
+	fill = image(icon, target, "fill")
+	fill.filters = filter(type = "alpha", icon = icon(icon, "mask"), x = -fill_width)
+
+	holder.overlays += backdrop
+	holder.overlays += fill
+
+/**
+  * Call periodically to update the progress bar's progress.
+  *
+  * Arguments:
+  * * progress - The new value, divided with the given goal on creation to calculate the fraction.
+  */
+/datum/progress_bar/proc/update(progress)
+	if(QDELETED(target))
+		qdel(src)
+		return
+
+	if(user?.client != client)
```

Comment:
This block is going to runtime or do a bunch of `null -= foo` when `client` is null, would be better to explicitly check it here and remove all the `?`.